### ISSUE TITLE
Remove dependency on mox3

### ISSUE THREAD

**Issue by mrbean-bremen** [jmcgeheeiv/pyfakefs#182]
See discussion in #179 - mox3 support is dropped by OpenStack as stated by @dmsimard, and mock is included in unittest since Python 3.3.
We may have a second go at this replacement (see #48 for the first try).

**Comment by mrbean-bremen** [comment]
I made a first version (mostly re-doing the previous changes) in https://github.com/mrbean-bremen/pyfakefs/tree/use-mock.
@jmcgeheeiv - what kind of tests would you suggest additionally to make sure that this works, or to reproduce the problems? I have seen the issue you mentioned about `unittest.main.os` not being patched, but I'm not sure if or why this is a problem for the library usage - can you give me an example?

**Comment by jmcgeheeiv** [comment]
I was just thinking this would need its own issue.

As I recall, the example I gave in [mock issue 250](https://github.com/testing-cabal/mock/issues/250) is not an issue for pyfakefs.  I believe I gave that example merely because it used a public file familiar to the `mock` maintainers. 

As I recall (and as described in [mock issue 250](https://github.com/testing-cabal/mock/issues/250)), the problem with `mock` is that it stores names instead of references.  So if thereafter the name is shadowed, you get something different when you look up the name.  `mox` does not have this problem, and it has performed flawlessly.

After I added my fabulous (to me, at least) port to `mock`, it was with great disappointment that I backed it out and returned to `mox`.  While I would be delighted to see my creation resurrected as you have done, if any module shadows a name that we are patching, as `unittest` did in the example, it will break pyfakefs.

Since `mox3` is winding down, I encourage you to simply fork the patch portion of `mox3` (and its tests) and add it to pyfakefs.  It works for us, and whatever patcher we use, it is completely internal to pyfakefs so client programmers will not notice the difference.

If you are looking for glory, fix `mock`.  That would make you a hero.

**Comment by mrbean-bremen** [comment]
Ok, using the stubber from mox3 sounds reasonable, if the mock problem is what you describe. I'm not from the stuff heroes are made of, so I think I will leave mock alone for the time being, and go the easy way. This may take some time, though...

**Comment by jmcgeheeiv** [comment]
This is a fairly slow moving event, so that is fine.  My impression is that the only files you need are `mox3/stubout.py` and the corresponding test.

**Comment by mrbean-bremen** [comment]
Ok, thanks - I may check this tomorrow, if I find the time.

**Comment by mrbean-bremen** [comment]
Ok - the stubout part that we need turned out to be surprisingly small and simple, and, other than the whole mox3, does not have any additional dependencies. I copied it over into pyfakes (will make a PR in a minute), and it works as expected.
I could not add the tests, as they depend on mox3, so I will check how to write tests for stubout. Also, I did not change the documentation in the file - I'm not sure what we are supposed to add if we use a part of another library (also under Apache license) - I would appreciate your help here.

**Comment by jmcgeheeiv** [comment]
I suggest leaving the inline documentation as-is, but not adding it to the pyfakefs documentation.

**Comment by mrbean-bremen** [comment]
Ok, thanks - I will do this another day with a clear head.

**Comment by mrbean-bremen** [comment]
I added a few tests - please check if this is sufficient.

**Comment by mrbean-bremen** [comment]
Fixed per #183.
